sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala

```diff
@@ -270,6 +271,18 @@ private[sql] object PartitioningUtils {
   private val upCastingOrder: Seq[DataType] =
     Seq(NullType, IntegerType, LongType, FloatType, DoubleType, StringType)
 
+  def checkPartitionColumnOfValidDataType(
+      schema: StructType,
+      partitionColumns: Array[String]): Unit = {
+
+    ResolvedDataSource.partitionColumnsSchema(schema, partitionColumns).foreach { field =>
+      field.dataType match {
+        case _: AtomicType | NullType => // OK
```

Comment:
I see. Just tried this snippet in `spark-shell`:

```
scala> import sqlContext.implicits._
import sqlContext.implicits._

scala> (0 until 10).map(_ => (null, null)).toDF("a", "b")
scala.MatchError: Null (of class scala.reflect.internal.Types$TypeRef$$anon$6)
        at org.apache.spark.sql.catalyst.ScalaReflection$class.schemaFor(ScalaReflection.scala:92)
        at org.apache.spark.sql.catalyst.ScalaReflection$.schemaFor(ScalaReflection.scala:29)
        at org.apache.spark.sql.catalyst.ScalaReflection$$anonfun$schemaFor$1.apply(ScalaReflection.scala:128)
        at org.apache.spark.sql.catalyst.ScalaReflection$$anonfun$schemaFor$1.apply(ScalaReflection.scala:126)
```

Seems that we simply don't match `Null` when inferring the schema. Also, the partition discovery process always promotes partition columns of `NullType` to `StringType` (e.g., you may have a partition column whose values are all `null`). This is similar to the JSON data source.